src/github.com/matrix-org/dendrite/roomserver/input/membership.go

```diff
@@ -0,0 +1,223 @@
+		}
+		if change.added.EventNID != 0 {
+			ev, _ := eventMap(events).lookup(change.added.EventNID)
+			if ev != nil {
+				ae = &ev.Event
+			}
+			targetNID = change.added.EventStateKeyNID
+		}
+		if updates, err = updateMembership(updater, targetNID, re, ae, updates); err != nil {
+			return nil, err
+		}
+	}
+	return nil, nil
+}
+
+func updateMembership(
+	updater types.RoomRecentEventsUpdater, targetNID types.EventStateKeyNID,
+	remove *gomatrixserverlib.Event, add *gomatrixserverlib.Event,
+	updates []api.OutputEvent,
+) ([]api.OutputEvent, error) {
+	var err error
+	old := "leave"
+	new := "leave"
+
+	if remove != nil {
+		old, err = remove.Membership()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if add != nil {
+		new, err = add.Membership()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if old == new {
+		return updates, nil
+	}
+
+	mu, err := updater.MembershipUpdater(targetNID)
+	if err != nil {
+		return nil, err
+	}
+
+	return updateWithNewMembership(mu, new, add, updates)
+}
+
+func updateWithNewMembership(
+	mu types.MembershipUpdater, newMembership string, add *gomatrixserverlib.Event,
+	updates []api.OutputEvent,
+) ([]api.OutputEvent, error) {
+	switch newMembership {
+	case "invite":
+		needsSending, err := mu.SetToInvite(*add)
+		if err != nil {
+			return nil, err
+		}
+		if needsSending {
+			updates = appendInviteUpdate(updates, add)
+		}
+		return updates, nil
+	case "join":
+		if mu.IsJoin() {
+			return updates, nil
+		}
+		retired, err := mu.SetToJoin(add.Sender())
+		if err != nil {
+			return nil, err
+		}
+		return appendRetireUpdates(updates, add, newMembership, retired), nil
+	case "leave", "ban":
+		if mu.IsLeave() {
+			return updates, nil
+		}
+		retired, err := mu.SetToLeave(add.Sender())
+		if err != nil {
+			mu.SetToLeave(add.Sender())
+		}
+		return appendRetireUpdates(updates, add, newMembership, retired), nil
+	default:
+		panic(fmt.Errorf(
+			"input: membership %q is not one of the allowed values", newMembership,
+		))
+	}
+}
+
+func appendInviteUpdate(
+	updates []api.OutputEvent, add *gomatrixserverlib.Event,
+) []api.OutputEvent {
+	onie := api.OutputNewInviteEvent{
+		Event: *add,
+	}
+	return append(updates, api.OutputEvent{
+		Type:           api.OutputTypeNewInviteEvent,
+		NewInviteEvent: &onie,
+	})
+}
+
+func appendRetireUpdates(
+	updates []api.OutputEvent, add *gomatrixserverlib.Event, membership string, retired []string,
+) []api.OutputEvent {
+	for _, eventID := range retired {
+		orie := api.OutputRetireInviteEvent{
+			EventID:    eventID,
+			Membership: membership,
+		}
+		if add != nil {
+			orie.RetiredByEventID = add.EventID()
+		}
+		updates = append(updates, api.OutputEvent{
+			Type:              api.OutputTypeRetireInviteEvent,
+			RetireInviteEvent: &orie,
+		})
+	}
+	return updates
+}
+
+type stateChange struct {
+	removed types.StateEntry
+	added   types.StateEntry
+}
+
+func pairUpChanges(removed, added []types.StateEntry) []stateChange {
```

Comment:
Not a great function name. This is grouping together state entries based on their state key tuple. I'd prefer `groupByStateKey(removed, added)` personally.